File: LeetCode_Solution/LeetCode_507/Solution.java
/**
 *  Problem Link- https://leetcode.com/problems/perfect-number/
 * 
 */

package LeetCode_Solution.LeetCode_507;

public class Solution {

  //! Brute Force Approach
  /**
   * Checks if a given number is a perfect number using a brute force approach.
   * 
   * Time Complexity: O(n)
   * Space Complexity: O(1)
   * 
   * @param num The number to be checked.
   * @return True if the number is a perfect number, false otherwise.
   */
  public static boolean checkPerfectNumberBruteForce(int num){
    int sum=0;
    for(int i=1;i<=num;i++){
      if(num%i==0 & i!=num) sum+=i;
    }
    return sum==num;
  }


  //! Optimize solution
  /**
   * Checks if a given number is a perfect number.
   * 
   * Time Complexity: O(sqrt(n))
   * Space Complexity: O(1)
   * 
   * @param num the number to be checked
   * @return true if the number is a perfect number, false otherwise
   */
  public static boolean checkPerfectNumber(int n) {
    int sum = 0;
    for (int i = 1; i * i <= n; i++) {
      if (n % i == 0) {
        sum += i;
        if ((n / i) != i)
          sum += (n / i);
      }
    }
    return sum == n ;
  }

  public static void main(String[] args) {

    System.out.println(checkPerfectNumber(28));
  }
}

File: LeetCode_Solution/LeetCode_507/Solution.js
/**
 *
 *  Problem Link: https://leetcode.com/problems/perfect-number/
 *
 */

//! BruteForce
/**
 * Checks if a number is a perfect number.
 *
 * TimeComplexity - O(n)
 * SpaceComplexity - O(1)
 *
 * @param {number} num - The number to be checked.
 * @returns {boolean} - Returns true if the number is a perfect number, otherwise returns false.
 */
let checkPerfectNumberBruteForce = (num) => {
  let sum = 0;
  for (let i = 1; i < num; i++) {
    if (num % i == 0 && i != num) sum += 0;
  }
  return sum == num;
};

//! Optimize Approach

/**
 * Checks if a number is a perfect number.
 * 
 * TimeComplexity - O(sqrt(n))
 * SpaceComplexity - O(1)
 *
 * @param {number} num - The number to be checked.
 * @returns {boolean} - Returns true if the number is a perfect number, otherwise returns false.
 */
var checkPerfectNumber = function (num) {
  let sum = 0;
  for (let i = 1; i * i <= num; i++) {
    if (num % i == 0 && i != num) {
      sum += i;
      let N = Math.floor(num / i);
      if (N != i && N != num) sum += N;
    }
  }
  return sum == num;
};

console.log(checkPerfectNumber(28)); // Output: true